lisa/nic.py

```diff
@@ -0,0 +1,222 @@
+
+        assert_that(self.is_empty()).is_false()
+
+    def _get_default_nic(self, nic_list: List[str]) -> str:
+        cmd = "/sbin/ip route | grep default | awk '{print $5}'"
+        default_interface = self._node.execute(cmd, shell=True, sudo=True)
+        default_interface.assert_exit_code()
+        assert_that(len(default_interface.stdout)).is_greater_than(0)
+        default_interface_name = default_interface.stdout.strip()
+        assert_that(default_interface_name in nic_list).described_as(
+            (
+                f"ERROR: NIC name found as default {default_interface_name} "
+                f"was not in original list of nics {nic_list}."
+            )
+        ).is_true()
+        return default_interface_name
+
+    def _get_host_if_info(self) -> None:
+        # for parsing ip addr show (ipv4)
+        _ip_regex = (
+            r"inet\s+([0-9a-fA-F]{1,3}\.[0-9a-fA-F]{1,3}\."
+            r"[0-9a-fA-F]{1,3}\.[0-9a-fA-F]{1,3})"
+        )
+        _mac_regex = (
+            r"ether\s+([0-9a-fA-F]{2}:[0-9a-fA-F]{1,3}:"
+            r"[0-9a-fA-F]{2}:[0-9a-fA-F]{2}:"
+            r"[0-9a-fA-F]{2}:[0-9a-fA-F]{2})"
+        )
+
+        for nic in self.get_upper_nics():
+            # get ip and mac
+            result = self._node.execute(f"/sbin/ip addr show {nic}", shell=True)
+            result.assert_exit_code()
+            ip_match = re.search(_ip_regex, result.stdout)
+            mac_match = re.search(_mac_regex, result.stdout)
+
+            if ip_match and mac_match:
+                # check we found matches for both
+                for match in [ip_match, mac_match]:
+                    assert_that(len(match.groups())).described_as(
+                        (
+                            f"(IP) Trouble parsing `ip addr show {nic}` output."
+                            " Number of match groups was unexpected."
+                        )
+                    ).is_equal_to(1)
+
+                ip_addr = ip_match.group(1)
+                mac_addr = mac_match.group(1)
+
+                # double check IP address looks right
+                try:
+                    ipaddress.ip_address(ip_addr)
+                except ValueError:
+                    assert_that(False).described_as(
+                        f"{ip_addr} was not recognized as an IP Address."
+                    ).is_true()
+
+                # save them both off
+                self.get_nic(nic)._ip_addr = ip_addr
+                self.get_nic(nic)._mac_addr = mac_addr
+            else:
+                assert_that(False).described_as(
+                    f"Could not parse output of ip addr show {nic}"
+                ).is_true()
+
+    def __init__(self, node: Node):
+        self._node = node
+        self._nics: Dict[str, NicInfo] = dict()
+        nics = self._get_nic_names()
+        self._get_node_nic_info(nics)
+        default_nic = self._get_default_nic(nics)
+        awk_cmd = "awk '{print $2}'"
```

Comment:
Don't execute any time cost method in `__init__` method. it's a construction method. Use `InitializableMixin` pattern, and initialize it with node.

Reply:
Ah, sorry. I guess this one slipped through before. Will fix.